File: tasks/builtin/refactor-03-go/starter.go
package solution

import (
	"fmt"
	"strconv"
	"strings"
)

func parseLines(rawInput string) [][]string {
	lines := strings.Split(strings.TrimSpace(rawInput), "\n")
	rows := [][]string{}
	for _, line := range lines {
		if strings.TrimSpace(line) != "" {
			parts := strings.Split(line, ",")
			rows = append(rows, parts)
		}
	}
	return rows
}

func validateRows(rows [][]string, expectedColumns int) ([][]string, []string) {
	validRows := [][]string{}
	errors := []string{}
	for i, row := range rows {
		if len(row) != expectedColumns {
			errors = append(errors, fmt.Sprintf("Row %d: expected %d columns, got %d", i, expectedColumns, len(row)))
		} else {
			validRows = append(validRows, row)
		}
	}
	return validRows, errors
}

func parseNumericColumns(rows [][]string, columnIndices map[int]bool) ([][]interface{}, []string) {
	parsedRows := [][]interface{}{}
	errors := []string{}
	for i, row := range rows {
		parsedRow := []interface{}{}
		hasError := false
		for j, cell := range row {
			if columnIndices[j] {
				val, err := strconv.ParseFloat(strings.TrimSpace(cell), 64)
				if err != nil {
					errors = append(errors, fmt.Sprintf("Row %d, column %d: '%s' is not a valid number", i, j, cell))
					hasError = true
					break
				}
				parsedRow = append(parsedRow, val)
			} else {
				parsedRow = append(parsedRow, strings.TrimSpace(cell))
			}
		}
		if !hasError {
			parsedRows = append(parsedRows, parsedRow)
		}
	}
	return parsedRows, errors
}

func normalizeTextColumns(rows [][]interface{}, columnIndices map[int]bool) [][]interface{} {
	normalizedRows := [][]interface{}{}
	for _, row := range rows {
		normalizedRow := []interface{}{}
		for j, cell := range row {
			if columnIndices[j] {
				if str, ok := cell.(string); ok {
					normalizedRow = append(normalizedRow, strings.ToUpper(str))
				} else {
					normalizedRow = append(normalizedRow, cell)
				}
			} else {
				normalizedRow = append(normalizedRow, cell)
			}
		}
		normalizedRows = append(normalizedRows, normalizedRow)
	}
	return normalizedRows
}

func computeDerivedColumn(rows [][]interface{}, sourceColIdx int, multiplier float64) [][]interface{} {
	resultRows := [][]interface{}{}
	for _, row := range rows {
		newRow := make([]interface{}, len(row)+1)
		copy(newRow, row)
		if val, ok := row[sourceColIdx].(float64); ok {
			newRow[len(row)] = val * multiplier
		}
		resultRows = append(resultRows, newRow)
	}
	return resultRows
}

func formatRowsAsTable(rows [][]interface{}, columnWidths []int) string {
	lines := []string{}
	for _, row := range rows {
		formattedCells := []string{}
		for i, cell := range row {
			width := columnWidths[i]
			if val, ok := cell.(float64); ok {
				formattedCells = append(formattedCells, fmt.Sprintf("%*.2f", width, val))
			} else {
				formattedCells = append(formattedCells, fmt.Sprintf("%-*s", width, fmt.Sprint(cell)))
			}
		}
		lines = append(lines, strings.Join(formattedCells, " | "))
	}
	return strings.Join(lines, "\n")
}

func ProcessData(rawInput string) string {
	rows := parseLines(rawInput)
	if len(rows) == 0 {
		return "ERROR: No data"
	}

	validRows, validationErrors := validateRows(rows, 3)
	if len(validationErrors) > 0 {
		return "ERROR: " + strings.Join(validationErrors, "; ")
	}

	parsedRows, parseErrors := parseNumericColumns(validRows, map[int]bool{2: true})
	if len(parseErrors) > 0 {
		return "ERROR: " + strings.Join(parseErrors, "; ")
	}

	normalizedRows := normalizeTextColumns(parsedRows, map[int]bool{0: true, 1: true})
	computedRows := computeDerivedColumn(normalizedRows, 2, 1.5)
	result := formatRowsAsTable(computedRows, []int{15, 12, 10, 10})

	return result
}

File: tasks/builtin/refactor-03-js/starter.js
function parseLines(rawInput) {
    const lines = rawInput.trim().split('\n');
    const rows = [];
    for (const line of lines) {
        if (line.trim()) {
            const parts = line.split(',');
            rows.push(parts);
        }
    }
    return rows;
}

function validateRows(rows, expectedColumns) {
    const validRows = [];
    const errors = [];
    for (let i = 0; i < rows.length; i++) {
        const row = rows[i];
        if (row.length !== expectedColumns) {
            errors.push(`Row ${i}: expected ${expectedColumns} columns, got ${row.length}`);
        } else {
            validRows.push(row);
        }
    }
    return { validRows, errors };
}

function parseNumericColumns(rows, columnIndices) {
    const parsedRows = [];
    const errors = [];
    for (let i = 0; i < rows.length; i++) {
        const row = rows[i];
        const parsedRow = [];
        let hasError = false;
        for (let j = 0; j < row.length; j++) {
            const cell = row[j];
            if (columnIndices.has(j)) {
                const num = parseFloat(cell.trim());
                if (isNaN(num)) {
                    errors.push(`Row ${i}, column ${j}: '${cell}' is not a valid number`);
                    hasError = true;
                    break;
                }
                parsedRow.push(num);
            } else {
                parsedRow.push(cell.trim());
            }
        }
        if (!hasError) {
            parsedRows.push(parsedRow);
        }
    }
    return { parsedRows, errors };
}

function normalizeTextColumns(rows, columnIndices) {
    const normalizedRows = [];
    for (const row of rows) {
        const normalizedRow = [];
        for (let j = 0; j < row.length; j++) {
            const cell = row[j];
            if (columnIndices.has(j) && typeof cell === 'string') {
                normalizedRow.push(cell.toUpperCase());
            } else {
                normalizedRow.push(cell);
            }
        }
        normalizedRows.push(normalizedRow);
    }
    return normalizedRows;
}

function computeDerivedColumn(rows, sourceColIdx, multiplier) {
    const resultRows = [];
    for (const row of rows) {
        const newRow = [...row, row[sourceColIdx] * multiplier];
        resultRows.push(newRow);
    }
    return resultRows;
}

function formatRowsAsTable(rows, columnWidths) {
    const lines = [];
    for (const row of rows) {
        const formattedCells = [];
        for (let i = 0; i < row.length; i++) {
            const cell = row[i];
            const width = columnWidths[i];
            if (typeof cell === 'number') {
                formattedCells.push(cell.toFixed(2).padStart(width));
            } else {
                formattedCells.push(String(cell).padEnd(width));
            }
        }
        lines.push(formattedCells.join(' | '));
    }
    return lines.join('\n');
}

function processData(rawInput) {
    const rows = parseLines(rawInput);
    if (rows.length === 0) {
        return 'ERROR: No data';
    }

    const { validRows, errors: validationErrors } = validateRows(rows, 3);
    if (validationErrors.length > 0) {
        return 'ERROR: ' + validationErrors.join('; ');
    }

    const { parsedRows, errors: parseErrors } = parseNumericColumns(validRows, new Set([2]));
    if (parseErrors.length > 0) {
        return 'ERROR: ' + parseErrors.join('; ');
    }

    const normalizedRows = normalizeTextColumns(parsedRows, new Set([0, 1]));
    const computedRows = computeDerivedColumn(normalizedRows, 2, 1.5);
    const result = formatRowsAsTable(computedRows, [15, 12, 10, 10]);

    return result;
}

module.exports = { processData };
